src/idpi/operators/lateral_operators.py

```diff
@@ -0,0 +1,192 @@
+    window_type : {"exp", "const"}
+        The window type.
+
+        Supported types:
+         - exp: Exponential drop from the center of the window.
+         - const: All weights are set to 1.
+
+    window_shape : {"disk", "square"}
+        The window shape.
+
+        Supported shapes:
+         - disk: The kernel is confined to the circle inscribing the window.
+         - square: The kernel is defined on the entire window.
+
+    Raises
+    ------
+    ValueError
+        if `window_size` is not an odd number.
+
+    Returns
+    -------
+    xarray.DataArray
+        Weights corresponding to the `window_type` within the given `window_shape`
+        and zero outside of the shape.
+
+    """
+    n = window_size
+
+    if n % 2 != 1:
+        raise ValueError("window_size must be an odd number.")
+
+    radius = n // 2
+    yy, xx = np.mgrid[:n, :n] - radius
+    dist = np.sqrt(xx**2 + yy**2)
+
+    if window_type == "exp":
+        kernel = np.exp(-dist)
+    else:
+        kernel = np.ones((n, n))
+
+    weights = xr.DataArray(kernel, dims=["win_x", "win_y"])
+
+    if window_shape == "disk":
+        return weights.where(dist <= radius, 0.0)
+
+    return weights
+
+
+def compute_cond_mask(
+    windows: xr.DataArray, weights: xr.DataArray, frac_val: float, nx: int, ny: int
+) -> xr.DataArray:
+    """Compute the conditional mask for which the convolution results are valid.
+
+    The function computes the ratio of undefined values to the total number values
+    in the convolution window. The computational domain and the window shape are both
+    taken into account. The mask is true for locations where the ratio is above the
+    given threshold.
+
+    Parameters
+    ----------
+    windows : xarray.DataArray
+        Values of the constructed windows for the application of the convolution.
+    weights : xarray.DataArray
+        Weights of the convolution kernel.
+    frac_val : float
+        Threshold on the ratio of undefined values.
+    nx : int
+        Size of the field in the x dimension.
+    ny : int
+        Size of the field in the y dimension.
+
+    Returns
+    -------
+    xarray.DataArray
+        Boolean mask indicating where the undefined value ratio in the convolution
+        kernel is within the acceptable threshold.
+
+    """
+    mask = weights > 0
+
+    loc_x = windows.x + windows.win_x - windows.sizes["win_x"] // 2
+    in_bnds_x = np.logical_and(0 <= loc_x, loc_x < nx)
+
+    loc_y = windows.y + windows.win_y - windows.sizes["win_y"] // 2
+    in_bnds_y = np.logical_and(0 <= loc_y, loc_y < ny)
+
+    undef = windows.isnull()
+    frac_undef = undef.where(mask).where(in_bnds_x).where(in_bnds_y).mean(weights.dims)
+    return frac_undef <= frac_val
+
+
+def fill_undef(field: xr.DataArray, radius: int, frac_val: float) -> xr.DataArray:
+    """Fill undefined values.
+
+    Replace undefined values with a value derived from neighbourhood.
```

Comment:
can you specify in the doc how the value from the neighbourhood is computed?

Reply:
It's essentially a Gaussian filter with sigma = 1.